crates/matrix-sdk-crypto/src/olm/group_sessions/sender_data_finder.rs

```diff
@@ -0,0 +1,1022 @@
+///                                     ┌─────────────────────────────────┘
+///                                     ▼
+/// ┌───────────────────────────────────────────────────────────────────┐
+/// │ H (cross-signing key matches that used to sign the device info!)  │
+/// └───────────────────────────────────────────────────────────────────┘
+///                                     │
+///   __________________________________▼______________________________
+///  ╱                                                                 ╲
+/// ╱ Is the signature in the device info valid?                        ╲yes
+/// ╲___________________________________________________________________╱ │
+///                                     │ no                              │
+///                                     ▼                                 │
+/// ╭───────────────────────────────────────────────────────────────────╮ │
+/// │ !Session is invalid: drop it from the store and forget it.        │ │
+/// ╰───────────────────────────────────────────────────────────────────╯ │
+///                                     ┌─────────────────────────────────┘
+///                                     ▼
+/// ╭───────────────────────────────────────────────────────────────────╮
+/// │ J (device info is verified by matching cross-signing key)         │
+/// │                                                                   │
+/// │ Look up the user_id and master_key for the user sending the       │
+/// │ to-device message.                                                │
+/// │                                                                   │
+/// │ Decide the master_key trust level based on whether we have        │
+/// │ verified this user.                                               │
+/// │                                                                   │
+/// │ Store this information with the session.                          │
+/// ╰───────────────────────────────────────────────────────────────────╯
+///
+/// Note: the sender data may become out-of-date if we later verify the user. We
+/// have no plans to update it if so.
+pub(crate) struct SenderDataFinder<'a> {
+    store: &'a Store,
+}
+
+impl<'a> SenderDataFinder<'a> {
+    /// As described in https://github.com/matrix-org/matrix-rust-sdk/issues/3543
+    /// and https://github.com/matrix-org/matrix-rust-sdk/issues/3544
+    /// find the device info associated with the to-device message used to
+    /// create the InboundGroupSession we are about to create, and decide
+    /// whether we trust the sender.
+    pub(crate) async fn find_using_event(
+        store: &'a Store,
+        sender_curve_key: Curve25519PublicKey,
+        room_key_event: &'a DecryptedRoomKeyEvent,
+    ) -> OlmResult<SenderData> {
+        let finder = Self { store };
+        finder.have_event(sender_curve_key, room_key_event).await
+    }
+
+    /// As described in https://github.com/matrix-org/matrix-rust-sdk/issues/3543
```

Comment:
Again links to issues, why can't we reference the flowchart in this file instead?